src/EventStore.Client/GossipChannelSelector.cs

```diff
@@ -0,0 +1,213 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+#nullable enable
+namespace EventStore.Client {
+	internal class GossipChannelSelector : IChannelSelector {
+		private readonly EventStoreClientSettings _settings;
+		private readonly IServerCapabilitiesClient _serverCapabilities;
+		private readonly CancellationToken _cancellationToken;
+
+		private static readonly ClusterMessages.VNodeState[] NotAllowedStates = {
+			ClusterMessages.VNodeState.Manager,
+			ClusterMessages.VNodeState.ShuttingDown,
+			ClusterMessages.VNodeState.Shutdown,
+			ClusterMessages.VNodeState.Unknown,
+			ClusterMessages.VNodeState.Initializing,
+			ClusterMessages.VNodeState.CatchingUp,
+			ClusterMessages.VNodeState.ResigningLeader,
+			ClusterMessages.VNodeState.PreLeader,
+			ClusterMessages.VNodeState.PreReplica,
+			ClusterMessages.VNodeState.Clone,
+			ClusterMessages.VNodeState.DiscoverLeader
+		};
+
+		private readonly Dictionary<EndPoint, ChannelBase> _grpcChannelsCache;
+		private readonly Dictionary<EndPoint, ServerCapabilities> _serverCapabilitiesCache;
+		private readonly System.Threading.Channels.Channel<ChannelInfo> _channelInfoChannel;
+		private readonly System.Threading.Channels.Channel<ControlMessage> _controlChannel;
+		private readonly ILogger<GossipChannelSelector> _log;
+		private int _disposed;
+
+		public GossipChannelSelector(EventStoreClientSettings settings, IGossipClient gossipClient,
+			IServerCapabilitiesClient serverCapabilities, CancellationToken cancellationToken = default) {
+			_settings = settings;
+			_serverCapabilities = serverCapabilities;
+			_cancellationToken = cancellationToken;
+			_grpcChannelsCache = new Dictionary<EndPoint, ChannelBase>();
+			_serverCapabilitiesCache = new Dictionary<EndPoint, ServerCapabilities>();
+			_channelInfoChannel = System.Threading.Channels.Channel.CreateUnbounded<ChannelInfo>();
+			_controlChannel = System.Threading.Channels.Channel.CreateUnbounded<ControlMessage>();
+			_log = settings.LoggerFactory?.CreateLogger<GossipChannelSelector>() ??
+			       new NullLogger<GossipChannelSelector>();
+
+			foreach (var endPoint in settings.ConnectivitySettings.GossipSeeds) {
+				_grpcChannelsCache.Add(endPoint, ChannelFactory.CreateChannel(settings, endPoint));
+			}
+
+			_controlChannel.Writer.TryWrite(ControlMessage.Discover.Instance);
+
+			_ = Task.Run(async () => {
```

Comment:
there's quite a lot going on in this lambda, might be worth making it a first class method and perhaps separating the main pump from the part that handles the discovery control message